### ISSUE TITLE
espeak (uses portaudio) won't speak with alsa-lib 1.2.5

### ISSUE THREAD

**Issue by zqqw** [PortAudio/portaudio#599]
This looks to be possibly due to an issue with portaudio:
https://forum.artixlinux.org/index.php/topic,2733.0.html
https://bbs.archlinux.org/viewtopic.php?id=267124
```
$ espeak
hello
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
attempt to connect to server failed
Expression 'framesAvail' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 3644
Expression 'PaAlsaStreamComponent_GetAvailableFrames( self, &framesAvail, xrun )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 4004
Expression 'PaAlsaStreamComponent_RegisterChannels( &self->playback, &self->bufferProcessor, &playbackFrames, &xrun )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 4133
Expression 'PaAlsaStream_SetUpBuffers( stream, &framesGot, &xrun )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 4374
hi
Expression 'framesAvail' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 3644
Expression 'PaAlsaStreamComponent_GetAvailableFrames( self, &framesAvail, xrun )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 4004
Expression 'PaAlsaStreamComponent_RegisterChannels( &self->playback, &self->bufferProcessor, &playbackFrames, &xrun )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 4133
Expression 'PaAlsaStream_SetUpBuffers( stream, &framesGot, &xrun )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 4374

After downgrading alsa-lib & lib32-alsa-lib to 1.2.4* it works, no reboot required :

$ espeak
hello
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
attempt to connect to server failed
hi
hola
```

**Comment by zqqw** [comment]
Related issue in alsa-lib:
https://github.com/alsa-project/alsa-lib/issues/155

**Comment by zqqw** [comment]
Bisecting alsa-lib reveals this to be the initial breaking commit, building on the previous commit works, on this one it doesn't. The error messages are different on some subsequent builds compared to the current state so there may be another later commit that affects things too.
```
commit 5275d170e05639473e68e5feb349aec65bf26428
Author: Alex Henrie <alexhenrie24@gmail.com>
Date:   Sat Dec 26 14:35:39 2020 -0700

    conf: fix use after free in _snd_config_load_with_include
    
    Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
    Signed-off-by: Takashi Iwai <tiwai@suse.de>
```

**Comment by philburk** [comment]
@zqqw - Thanks for investigating this. Do you know if ALSA 1.2.5.1 release fixes this problem?

Do you think this is purely a bug in ALSA or was it a legal change in ALSA that exposed a bug in PortAudio?

**Comment by RossBencina** [comment]
Searching the web for `PaAlsaStreamComponent_GetAvailableFrames` brings up quite a few discussions, for example:

https://github.com/Arkq/bluez-alsa/issues/243

https://forums.gentoo.org/viewtopic-t-1109266.html

**Comment by zqqw** [comment]
Building lib-alsa at the latest commit made yesterday does not work:
```
commit 81e7923fbfad45b2f353a4d6e3053af51f5f7d0b (HEAD -> master, origin/master, origin/HEAD)
Author: Jaroslav Kysela <perex@perex.cz>
Date:   Tue Jun 15 23:21:42 2021 +0200
```
Also reverting the commit that initially stops it working in the latest lib-alsa does not restore operation, so there are subsequent problems too. I have no idea at this point where the bug could be said to be located. But just looking at that commit mentioned above, I think it's reading alsa.conf in that file, and possibly that part of the bug is because input.current is used to set fd from the result of getchar before errno is checked.

**Comment by zqqw** [comment]
Bisecting again with the first faulty commit reverted reveals this as the next breaking point:
https://github.com/alsa-project/alsa-lib/commit/28cc099d9ea3962b033cb1cb9c3e07db828d9ff7
So far it's looking more like an alsa-lib bug.

**Comment by philburk** [comment]
Are the alsa-lib maintainers aware of this bug?

**Comment by zqqw** [comment]
Yes, hopefully they will look into it, and it's only a small amount of code changes causing the problem:
https://github.com/alsa-project/alsa-lib/issues/155

**Comment by philburk** [comment]
Let's leave this Issue open until the ALSA bug is fixed. That will make it easier for folks to find this Issue.

**Comment by zqqw** [comment]
Apologies for not updating this more promptly. It has been fixed by this commit:
https://github.com/alsa-project/alsa-lib/commit/dd609ef9684987d3ca61d5c5cc3c77589ff9c29f
For Arch & Arch based distros, building alsa-lib-git should include the fix:
 https://aur.archlinux.org/packages/alsa-lib-git/
But it may take some time for that to be included in the next release as 1.2.5 only recently came out, unless it's backported as a minor revision, I'm not sure how alsa-lib do things.

**Comment by philburk** [comment]
Fixed in ALSA.
